# Assembly Finder

[![version](https://img.shields.io/conda/v/bioconda/assembly_finder?label=version)](http://bioconda.github.io/recipes/assembly_finder/README.html)
[![downloads](https://img.shields.io/conda/dn/bioconda/assembly_finder)](https://anaconda.org/bioconda/assembly_finder)

Assembly finder is a Snakemake cli to download genomes with [NCBI datasets](https://github.com/ncbi/datasets).

## Installation

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/assembly_finder/README.html)

```sh
mamba create -n assembly_finder assembly_finder
```

## Usage

```sh
assembly_finder -i <input>
```

### Quick usage

```sh
assembly_finder -i taxons.txt
```

or

```sh
assembly_finder -i 1290,staphylococcus_aureus,287 -nb 1 -o taxons
```

## Output

This is how the taxons download directory looks like

```sh
ðŸ“‚taxons
 â”£ ðŸ“‚download
 â”ƒ â”£ ðŸ“‚GCF_000233495.1
 â”ƒ â”ƒ â”£ ðŸ“œGCF_000233495.1_Pseudomonas_sp_2_1_26_V1_genomic.fna.gz
 â”ƒ â”ƒ â”— ðŸ“œsequence_report.jsonl
 â”ƒ â”£ ðŸ“‚GCF_000418345.1
 â”ƒ â”ƒ â”£ ðŸ“œGCF_000418345.1_ASM41834v1_genomic.fna.gz
 â”ƒ â”ƒ â”— ðŸ“œsequence_report.jsonl
 â”ƒ â”£ ðŸ“‚GCF_003812505.1
 â”ƒ â”ƒ â”£ ðŸ“œGCF_003812505.1_ASM381250v1_genomic.fna.gz
 â”ƒ â”ƒ â”— ðŸ“œsequence_report.jsonl
 â”ƒ â”£ ðŸ“œ.snakemake_timestamp
 â”ƒ â”£ ðŸ“œassembly_data_report.jsonl
 â”ƒ â”— ðŸ“œdataset_catalog.json
 â”£ ðŸ“œarchive.zip
 â”£ ðŸ“œassembly_summary.tsv
 â”£ ðŸ“œsequence_report.tsv
 â”— ðŸ“œtaxonomy.tsv
```

## Examples

### Download the top 1 assemblies for each bacterial species

```sh
assembly_finder -i bacteria --api-key <api-key> --rank species --nrank 1
```

### Donwload all refseq bacteria viruses and archaea complete genomes (exclude MAGs and anomalous genomes)

```sh
assembly_finder -i bacteria,viruses,archaea -o outdir --api-key <api-key> --assembly-level complete --exclude-atypical --mag exclude
```

### Download specific genomes with accessions

```sh
assembly_finder -i GCF_003812505.1,GCF_000418345.1,GCF_000233495.1 -o accessions
```

## Parameters

```sh

 Usage: assembly_finder [OPTIONS] [SNAKEMAKE_ARGS]...


  â–‘â–ˆâ–€â–ˆâ–‘â–ˆâ–€â–€â–‘â–ˆâ–€â–€â–‘â–ˆâ–€â–€â–‘â–ˆâ–„â–ˆâ–‘â–ˆâ–€â–„â–‘â–ˆâ–‘â–‘â–‘â–ˆâ–‘â–ˆâ–‘â–‘â–‘â–ˆâ–€â–€â–‘â–€â–ˆâ–€â–‘â–ˆâ–€â–ˆâ–‘â–ˆâ–€â–„â–‘â–ˆâ–€â–€â–‘â–ˆâ–€â–„
  â–‘â–ˆâ–€â–ˆâ–‘â–€â–€â–ˆâ–‘â–€â–€â–ˆâ–‘â–ˆâ–€â–€â–‘â–ˆâ–‘â–ˆâ–‘â–ˆâ–€â–„â–‘â–ˆâ–‘â–‘â–‘â–‘â–ˆâ–‘â–‘â–‘â–‘â–ˆâ–€â–€â–‘â–‘â–ˆâ–‘â–‘â–ˆâ–‘â–ˆâ–‘â–ˆâ–‘â–ˆâ–‘â–ˆâ–€â–€â–‘â–ˆâ–€â–„
  â–‘â–€â–‘â–€â–‘â–€â–€â–€â–‘â–€â–€â–€â–‘â–€â–€â–€â–‘â–€â–‘â–€â–‘â–€â–€â–‘â–‘â–€â–€â–€â–‘â–‘â–€â–‘â–‘â–‘â–‘â–€â–‘â–‘â–‘â–€â–€â–€â–‘â–€â–‘â–€â–‘â–€â–€â–‘â–‘â–€â–€â–€â–‘â–€â–‘â–€

 Snakemake-cli to download genomes with NCBI datasets

â•­â”€ Options â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚ *  --input    -i   TEXT                            path to assembly_finder input   â”‚
â”‚                                                    table or list of queries        â”‚
â”‚                                                    [required]                      â”‚
â”‚    --outdir   -o   PATH                            output directory                â”‚
â”‚    --number   -nb  INTEGER                         number of assemblies per query  â”‚
â”‚    --threads  -t   INTEGER                         number of threads to allow for  â”‚
â”‚                                                    the workflow                    â”‚
â”‚                                                    [default: 2]                    â”‚
â”‚    --taxon                                         are inputs taxon names or ids   â”‚
â”‚                                                    [default: True]                 â”‚
â”‚    --rank          [superkingdom|phylum|class|ord  taxonomic rank to filter by     â”‚
â”‚                    er|family|genus|species]        assemblies                      â”‚
â”‚    --nrank         INTEGER                         number of genomes per taxonomic â”‚
â”‚                                                    rank                            â”‚
â”‚    --dryrun                                        snakemake dryrun to see the     â”‚
â”‚                                                    scheduling plan                 â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
â•­â”€ NCBI datasets options â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚ --api-key           TEXT                  NCBI api-key                             â”‚
â”‚ --compressed                              Download compressed files                â”‚
â”‚ --source            [refseq|genbank|all]  download from refseq or genbank          â”‚
â”‚                                           [default: all]                           â”‚
â”‚ --include           TEXT                  Comma seperated files to download :      â”‚
â”‚                                           genome,rna,protein,cds,gff3,gtf,gbff,seâ€¦ â”‚
â”‚                                           [default: genome,seq-report]             â”‚
â”‚ --reference                               limit to reference and representative    â”‚
â”‚                                           genomes                                  â”‚
â”‚                                           [default: False]                         â”‚
â”‚ --assembly-level    TEXT                  select complete, chromosome, scaffold,   â”‚
â”‚                                           contig                                   â”‚
â”‚ --annotated                               select assemblies with annotation        â”‚
â”‚                                           [default: False]                         â”‚
â”‚ --atypical                                exclude atypical genomes [default: True] â”‚
â”‚ --mag               [only|exclude|all]    exclude or add MAGs to the dwnloaded     â”‚
â”‚                                           genomes                                  â”‚
â”‚                                           [default: all]                           â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
â•­â”€ Help â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚ --help     -h    Show this message and exit.                                       â”‚
â”‚ --version  -v    Show the version and exit.                                        â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
```
