## Download tables 

Starting from [v0.8.0](https://github.com/metagenlab/assembly_finder/releases/tag/v0.8.0), you can restrict outputs to `assembly_summary.tsv` and `taxonomy.tsv`

* Command

```sh
assembly_finder -i staphylococcus_aureus -nb 1 --summary
```

* Output

```sh
ğŸ“‚staphylococcus_aureus
 â”£ ğŸ“‚logs
 â”ƒ â”£ ğŸ“‚taxons
 â”ƒ â”ƒ â”— ğŸ“œstaphylococcus_aureus.log
 â”ƒ â”—ğŸ“œlineage.log
 â”£ ğŸ“œassembly_finder.log
 â”£ ğŸ“œassembly_summary.tsv
 â”£ ğŸ“œconfig.yaml
 â”— ğŸ“œtaxonomy.tsv
```

## Download genomes
### Small datasets
* *Staphylococcus aureus* reference genome

```sh
assembly_finder -i staphylococcus_aureus -nb 1
```
!!! note
    By default, assembly_finder limits genomes to reference or representative

* Any *Staphylococcus aureus* genome

```sh
assembly_finder -i staphylococcus_aureus -nb 1 --reference false
```

* Download from a list of taxa

```sh
assembly_finder -i 1290,1813735,114185 -o taxa -nb 1
```

* Download using a taxa table

??? info "**taxa.tsv**"
        

    | taxon | nb |
    | :-------------------- | :-- |
    | 1290 | 1 |
    | 1813735 | 1 |
    | 114185 | 1 |

```sh
assembly_finder -i taxa.tsv
```

### Big datasets

!!! warning

    These examples are for big datasets downloads, so using an NCBI api-key is highly recommended

* Download all chlamydia genomes

```sh
assembly_finder -i chlamydia --api-key <api-key>
```

* Best ranking genome for each bacteria species

```sh
assembly_finder -i bacteria --api-key <api-key> --rank species --nrank 1
```

* Complete RefSeq bacteria viruses and archaea <small>(excluding MAGs and atypical)</small>

```sh
assembly_finder -i bacteria,viruses,archaea --api-key <api-key> \
--source refseq --assembly-level complete --mag exclude --atypical \
-o outdir
```

* Specific bioproject

```sh
assembly_finder -i PRJNA289059 --api-key <api-key> --accession
```
## Download other files <small>(cds, proteins, gff3 ...)</small>
```sh
assembly_finder -i staphylococcus_aureus -nb 1 \
--include genome,rna,protein,cds,gff3,gtf,gbff,seq-report
```
Output:
```sh
ğŸ“‚staphylococcus_aureus
 â”£ ğŸ“‚download
 â”ƒ â”£ ğŸ“‚GCF_000013425.1
 â”ƒ â”ƒ â”£ ğŸ“œGCF_000013425.1_ASM1342v1_genomic.fna.gz
 â”ƒ â”ƒ â”£ ğŸ“œcds_from_genomic.fna.gz
 â”ƒ â”ƒ â”£ ğŸ“œgenomic.gbff.gz
 â”ƒ â”ƒ â”£ ğŸ“œgenomic.gff.gz
 â”ƒ â”ƒ â”£ ğŸ“œgenomic.gtf.gz
 â”ƒ â”ƒ â”— ğŸ“œprotein.faa.gz
 â”ƒ â”ƒ â”— ğŸ“œsequence_report.jsonl 
 â”ƒ â”— ğŸ“œ.snakemake_timestamp
 â”£ ğŸ“‚logs
 â”ƒ â”£ ğŸ“‚taxons
 â”ƒ â”ƒ â”— ğŸ“œstaphylococcus_aureus.log
 â”ƒ â”£ ğŸ“œarchive.log
 â”ƒ â”£ ğŸ“œlineage.log
 â”ƒ â”£ ğŸ“œrsync.log
 â”ƒ â”— ğŸ“œunzip.log
 â”£ ğŸ“œarchive.zip
 â”£ ğŸ“œassembly_finder.log
 â”£ ğŸ“œassembly_summary.tsv
 â”£ ğŸ“œconfig.yaml
 â”— ğŸ“œtaxonomy.tsv
```
