
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Farid Chaabane">
      
      
      
        <link rel="prev" href="../inputs/">
      
      
        <link rel="next" href="../examples/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.16">
    
    
      
        <title>Outputs - assembly finder</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.7e37652d.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-directory" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="assembly finder" class="md-header__button md-logo" aria-label="assembly finder" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            assembly finder
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Outputs
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/metagenlab/assembly_finder" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    metagenlab/assembly_finder
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="assembly finder" class="md-nav__button md-logo" aria-label="assembly finder" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    assembly finder
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/metagenlab/assembly_finder" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.0.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path fill="currentColor" d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    metagenlab/assembly_finder
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../inputs/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Inputs
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Outputs
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-directory" class="md-nav__link">
    <span class="md-ellipsis">
      Download directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-summary" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      Taxonomy
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Examples
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-directory" class="md-nav__link">
    <span class="md-ellipsis">
      Download directory
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-summary" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly summary
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      Taxonomy
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Outputs</h1>

<p>Below are all the outputs when using the <a href="../inputs/#tables">taxons table example</a>
<div class="language-sh highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>ðŸ“‚taxons
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚download
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚GCF_000008865.2
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œGCF_000008865.2_ASM886v2_genomic.fna.gz
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚GCF_000013425.1
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œGCF_000013425.1_ASM1342v1_genomic.fna.gz
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚GCF_003812505.1
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œGCF_003812505.1_ASM381250v1_genomic.fna.gz
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œ.snakemake_timestamp
</span><span id="__span-0-10"><a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚logs
</span><span id="__span-0-11"><a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“‚taxons
</span><span id="__span-0-12"><a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“œ1290.log
</span><span id="__span-0-13"><a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“œ562.log
</span><span id="__span-0-14"><a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œstaphylococcus_aureus.log
</span><span id="__span-0-15"><a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“œarchive.log
</span><span id="__span-0-16"><a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“œlineage.log
</span><span id="__span-0-17"><a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”£<span class="w"> </span>ðŸ“œrsync.log
</span><span id="__span-0-18"><a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="w"> </span>â”ƒ<span class="w"> </span>â”—<span class="w"> </span>ðŸ“œunzip.log
</span><span id="__span-0-19"><a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“œarchive.zip
</span><span id="__span-0-20"><a id="__codelineno-0-20" name="__codelineno-0-20" href="#__codelineno-0-20"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“œassembly_finder.log
</span><span id="__span-0-21"><a id="__codelineno-0-21" name="__codelineno-0-21" href="#__codelineno-0-21"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“œassembly_summary.tsv
</span><span id="__span-0-22"><a id="__codelineno-0-22" name="__codelineno-0-22" href="#__codelineno-0-22"></a><span class="w"> </span>â”£<span class="w"> </span>ðŸ“œconfig.yaml
</span><span id="__span-0-23"><a id="__codelineno-0-23" name="__codelineno-0-23" href="#__codelineno-0-23"></a><span class="w"> </span>â”—<span class="w"> </span>ðŸ“œtaxonomy.tsv
</span></code></pre></div></p>
<h2 id="download-directory">Download directory</h2>
<p>Downloaded files (<code>genomic.fna.gz</code> by default) are located in the download directory as shown below
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a> â”£ ðŸ“‚download
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a> â”ƒ â”£ ðŸ“‚GCF_000008865.2
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a> â”ƒ â”ƒ â”— ðŸ“œGCF_000008865.2_ASM886v2_genomic.fna.gz
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a> â”ƒ â”£ ðŸ“‚GCF_000013425.1
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a> â”ƒ â”ƒ â”— ðŸ“œGCF_000013425.1_ASM1342v1_genomic.fna.gz
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a> â”ƒ â”£ ðŸ“‚GCF_003812505.1
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a> â”ƒ â”ƒ â”— ðŸ“œGCF_003812505.1_ASM381250v1_genomic.fna.gz
</span></code></pre></div></p>
<h2 id="assembly-summary">Assembly summary</h2>
<p>Table with assembly informations such as assembly level, reference category, checkM and BUSCO completeness, sequencing technology, number of contigs ...</p>
<table>
<thead>
<tr>
<th style="text-align: left;">taxon</th>
<th style="text-align: left;">accession</th>
<th style="text-align: left;">current_accession</th>
<th style="text-align: left;">paired_accession</th>
<th style="text-align: left;">source_database</th>
<th style="text-align: left;">annotation_info.method</th>
<th style="text-align: left;">annotation_info.name</th>
<th style="text-align: left;">annotation_info.pipeline</th>
<th style="text-align: left;">annotation_info.provider</th>
<th style="text-align: left;">annotation_info.release_date</th>
<th style="text-align: left;">annotation_info.software_version</th>
<th style="text-align: left;">annotation_info.stats.gene_counts.non_coding</th>
<th style="text-align: left;">annotation_info.stats.gene_counts.protein_coding</th>
<th style="text-align: left;">annotation_info.stats.gene_counts.pseudogene</th>
<th style="text-align: left;">annotation_info.stats.gene_counts.total</th>
<th style="text-align: left;">assembly_level</th>
<th style="text-align: left;">assembly_method</th>
<th style="text-align: left;">assembly_name</th>
<th style="text-align: left;">assembly_status</th>
<th style="text-align: left;">assembly_type</th>
<th style="text-align: left;">bioproject_accession</th>
<th style="text-align: left;">biosample.accession</th>
<th style="text-align: left;">biosample.bioprojects</th>
<th style="text-align: left;">biosample.description.organism_name</th>
<th style="text-align: left;">biosample.description.tax_id</th>
<th style="text-align: left;">biosample.description.title</th>
<th style="text-align: left;">biosample.last_updated</th>
<th style="text-align: left;">biosample.models</th>
<th style="text-align: left;">biosample.owner.contacts</th>
<th style="text-align: left;">biosample.owner.name</th>
<th style="text-align: left;">biosample.package</th>
<th style="text-align: left;">biosample.publication_date</th>
<th style="text-align: left;">biosample.status.status</th>
<th style="text-align: left;">biosample.status.when</th>
<th style="text-align: left;">biosample.submission_date</th>
<th style="text-align: left;">paired_assembly.accession</th>
<th style="text-align: left;">paired_assembly.annotation_name</th>
<th style="text-align: left;">paired_assembly.status</th>
<th style="text-align: left;">refseq_category</th>
<th style="text-align: left;">release_date</th>
<th style="text-align: left;">sequencing_tech</th>
<th style="text-align: left;">submitter</th>
<th style="text-align: left;">contig_l50</th>
<th style="text-align: left;">contig_n50</th>
<th style="text-align: left;">gc_count</th>
<th style="text-align: left;">gc_percent</th>
<th style="text-align: left;">genome_coverage</th>
<th style="text-align: left;">number_of_component_sequences</th>
<th style="text-align: left;">number_of_contigs</th>
<th style="text-align: left;">number_of_scaffolds</th>
<th style="text-align: left;">scaffold_l50</th>
<th style="text-align: left;">scaffold_n50</th>
<th style="text-align: left;">total_number_of_chromosomes</th>
<th style="text-align: left;">total_sequence_length</th>
<th style="text-align: left;">total_ungapped_length</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.ani</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.assembly</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.assembly_coverage</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.category</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.organism_name</th>
<th style="text-align: left;">average_nucleotide_identity.best_ani_match.type_assembly_coverage</th>
<th style="text-align: left;">average_nucleotide_identity.category</th>
<th style="text-align: left;">average_nucleotide_identity.comment</th>
<th style="text-align: left;">average_nucleotide_identity.match_status</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.ani</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.assembly</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.assembly_coverage</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.category</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.organism_name</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_ani_match.type_assembly_coverage</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_organism</th>
<th style="text-align: left;">average_nucleotide_identity.submitted_species</th>
<th style="text-align: left;">average_nucleotide_identity.taxonomy_check_status</th>
<th style="text-align: left;">checkm_info.checkm_marker_set</th>
<th style="text-align: left;">checkm_info.checkm_marker_set_rank</th>
<th style="text-align: left;">checkm_info.checkm_species_tax_id</th>
<th style="text-align: left;">checkm_info.checkm_version</th>
<th style="text-align: left;">checkm_info.completeness</th>
<th style="text-align: left;">checkm_info.completeness_percentile</th>
<th style="text-align: left;">checkm_info.contamination</th>
<th style="text-align: left;">infraspecific_names.strain</th>
<th style="text-align: left;">organism_name</th>
<th style="text-align: left;">tax_id</th>
<th style="text-align: left;">path</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">staphylococcus_aureus</td>
<td style="text-align: left;">GCF_000013425.1</td>
<td style="text-align: left;">GCF_000013425.1</td>
<td style="text-align: left;">GCA_000013425.1</td>
<td style="text-align: left;">SOURCE_DATABASE_REFSEQ</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">Annotation submitted by NCBI RefSeq</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">NCBI RefSeq</td>
<td style="text-align: left;">2016-08-03</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">75</td>
<td style="text-align: left;">2767</td>
<td style="text-align: left;">30</td>
<td style="text-align: left;">2872</td>
<td style="text-align: left;">Complete Genome</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">ASM1342v1</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">haploid</td>
<td style="text-align: left;">PRJNA237</td>
<td style="text-align: left;">SAMN02604235</td>
<td style="text-align: left;">[{'accession': 'PRJNA237'}]</td>
<td style="text-align: left;">Staphylococcus aureus subsp. aureus NCTC 8325</td>
<td style="text-align: left;">93061</td>
<td style="text-align: left;">Sample from Staphylococcus aureus subsp. aureus NCTC 8325</td>
<td style="text-align: left;">2015-05-18T13:21:01.110</td>
<td style="text-align: left;">['Generic']</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">NCBI</td>
<td style="text-align: left;">Generic.1.0</td>
<td style="text-align: left;">2014-01-30T15:13:19.920</td>
<td style="text-align: left;">live</td>
<td style="text-align: left;">2014-01-30T15:13:19.920</td>
<td style="text-align: left;">2014-01-30T15:13:19.920</td>
<td style="text-align: left;">GCA_000013425.1</td>
<td style="text-align: left;">Annotation submitted by University of Oklahoma Health Sciences Center</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">reference genome</td>
<td style="text-align: left;">2006-02-13</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">University of Oklahoma Health Sciences Center</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">2821361</td>
<td style="text-align: left;">927332</td>
<td style="text-align: left;">33</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">2821361</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">2821361</td>
<td style="text-align: left;">2821361</td>
<td style="text-align: left;">99.94</td>
<td style="text-align: left;">GCA_006094915.1</td>
<td style="text-align: left;">96.32</td>
<td style="text-align: left;">type</td>
<td style="text-align: left;">Staphylococcus aureus</td>
<td style="text-align: left;">97.66</td>
<td style="text-align: left;">category_na</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">species_match</td>
<td style="text-align: left;">99.94</td>
<td style="text-align: left;">GCA_006094915.1</td>
<td style="text-align: left;">96.32</td>
<td style="text-align: left;">type</td>
<td style="text-align: left;">Staphylococcus aureus</td>
<td style="text-align: left;">97.66</td>
<td style="text-align: left;">Staphylococcus aureus subsp. aureus NCTC 8325</td>
<td style="text-align: left;">Staphylococcus aureus</td>
<td style="text-align: left;">OK</td>
<td style="text-align: left;">Staphylococcus aureus</td>
<td style="text-align: left;">species</td>
<td style="text-align: left;">1280</td>
<td style="text-align: left;">v1.2.2</td>
<td style="text-align: left;">97.59</td>
<td style="text-align: left;">19.683367</td>
<td style="text-align: left;">0.39</td>
<td style="text-align: left;">NCTC 8325</td>
<td style="text-align: left;">Staphylococcus aureus subsp. aureus NCTC 8325</td>
<td style="text-align: left;">93061</td>
<td style="text-align: left;">/path/to/genome/GCF_000013425.1/GCF_000013425.1_ASM1342v1_genomic.fna.gz</td>
</tr>
<tr>
<td style="text-align: left;">562</td>
<td style="text-align: left;">GCF_000008865.2</td>
<td style="text-align: left;">GCF_000008865.2</td>
<td style="text-align: left;">GCA_000008865.2</td>
<td style="text-align: left;">SOURCE_DATABASE_REFSEQ</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">Annotation submitted by NCBI RefSeq</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">NCBI RefSeq</td>
<td style="text-align: left;">2021-02-12</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">126</td>
<td style="text-align: left;">5155</td>
<td style="text-align: left;">136</td>
<td style="text-align: left;">5417</td>
<td style="text-align: left;">Complete Genome</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">ASM886v2</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">haploid</td>
<td style="text-align: left;">PRJNA226</td>
<td style="text-align: left;">SAMN01911278</td>
<td style="text-align: left;">[{'accession': 'PRJNA226'}]</td>
<td style="text-align: left;">Escherichia coli O157:H7 str. Sakai</td>
<td style="text-align: left;">386585</td>
<td style="text-align: left;">Bacterial, clinical or host-associated sample for Escherichia coli O157:H7 str. SAKAI (EHEC)</td>
<td style="text-align: left;">2019-05-23T15:25:40.989</td>
<td style="text-align: left;">['Pathogen.ba-cl']</td>
<td style="text-align: left;">[{}]</td>
<td style="text-align: left;">ATCC</td>
<td style="text-align: left;">Pathogen.cl.1.0</td>
<td style="text-align: left;">2013-02-05T00:00:00.000</td>
<td style="text-align: left;">live</td>
<td style="text-align: left;">2014-11-20T09:44:57</td>
<td style="text-align: left;">2013-02-05T09:09:06.203</td>
<td style="text-align: left;">GCA_000008865.2</td>
<td style="text-align: left;">Annotation submitted by GIRC</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">reference genome</td>
<td style="text-align: left;">2018-06-08</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">GIRC</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">5498578</td>
<td style="text-align: left;">2824389</td>
<td style="text-align: left;">50.5</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">5498578</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">5594605</td>
<td style="text-align: left;">5594605</td>
<td style="text-align: left;">99.97</td>
<td style="text-align: left;">GCA_001281725.1</td>
<td style="text-align: left;">94.32</td>
<td style="text-align: left;">claderef</td>
<td style="text-align: left;">Escherichia coli</td>
<td style="text-align: left;">99.57</td>
<td style="text-align: left;">category_na</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">species_match</td>
<td style="text-align: left;">99.97</td>
<td style="text-align: left;">GCA_001281725.1</td>
<td style="text-align: left;">94.32</td>
<td style="text-align: left;">claderef</td>
<td style="text-align: left;">Escherichia coli</td>
<td style="text-align: left;">99.57</td>
<td style="text-align: left;">Escherichia coli O157:H7 str. Sakai</td>
<td style="text-align: left;">Escherichia coli</td>
<td style="text-align: left;">OK</td>
<td style="text-align: left;">Escherichia coli</td>
<td style="text-align: left;">species</td>
<td style="text-align: left;">562</td>
<td style="text-align: left;">v1.2.2</td>
<td style="text-align: left;">99.51</td>
<td style="text-align: left;">92.85564</td>
<td style="text-align: left;">0.15</td>
<td style="text-align: left;">Sakai substr. RIMD 0509952</td>
<td style="text-align: left;">Escherichia coli O157:H7 str. Sakai</td>
<td style="text-align: left;">386585</td>
<td style="text-align: left;">/path/to/genome/GCF_000008865.2/GCF_000008865.2_ASM886v2_genomic.fna.gz</td>
</tr>
<tr>
<td style="text-align: left;">1290</td>
<td style="text-align: left;">GCF_003812505.1</td>
<td style="text-align: left;">GCF_003812505.1</td>
<td style="text-align: left;">GCA_003812505.1</td>
<td style="text-align: left;">SOURCE_DATABASE_REFSEQ</td>
<td style="text-align: left;">Best-placed reference protein set; GeneMarkS-2+</td>
<td style="text-align: left;">GCF_003812505.1-RS_2024_03_28</td>
<td style="text-align: left;">NCBI Prokaryotic Genome Annotation Pipeline (PGAP)</td>
<td style="text-align: left;">NCBI RefSeq</td>
<td style="text-align: left;">2024-03-28</td>
<td style="text-align: left;">6.7</td>
<td style="text-align: left;">85</td>
<td style="text-align: left;">2142</td>
<td style="text-align: left;">35</td>
<td style="text-align: left;">2262</td>
<td style="text-align: left;">Complete Genome</td>
<td style="text-align: left;">SMRT v. 2.3.0, HGAP v. 3.0</td>
<td style="text-align: left;">ASM381250v1</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">haploid</td>
<td style="text-align: left;">PRJNA231221</td>
<td style="text-align: left;">SAMN10163251</td>
<td style="text-align: left;">[{'accession': 'PRJNA231221'}]</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">1290</td>
<td style="text-align: left;">Pathogen: clinical or host-associated sample from Staphylococcus hominis</td>
<td style="text-align: left;">2019-05-14T13:08:20.304</td>
<td style="text-align: left;">['Pathogen.cl']</td>
<td style="text-align: left;">[{}]</td>
<td style="text-align: left;">US Food and Drug Administration</td>
<td style="text-align: left;">Pathogen.cl.1.0</td>
<td style="text-align: left;">2018-10-02T00:00:00.000</td>
<td style="text-align: left;">live</td>
<td style="text-align: left;">2018-10-02T12:23:11.101</td>
<td style="text-align: left;">2018-10-02T12:23:11.100</td>
<td style="text-align: left;">GCA_003812505.1</td>
<td style="text-align: left;">NCBI Prokaryotic Genome Annotation Pipeline (PGAP)</td>
<td style="text-align: left;">current</td>
<td style="text-align: left;">representative genome</td>
<td style="text-align: left;">2018-11-21</td>
<td style="text-align: left;">PacBio; Illumina</td>
<td style="text-align: left;">US Food and Drug Administration</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">2220494</td>
<td style="text-align: left;">713682</td>
<td style="text-align: left;">31.5</td>
<td style="text-align: left;">19.6x</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">1</td>
<td style="text-align: left;">2220494</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">2257431</td>
<td style="text-align: left;">2257431</td>
<td style="text-align: left;">99.99</td>
<td style="text-align: left;">GCA_900458635.1</td>
<td style="text-align: left;">98.99</td>
<td style="text-align: left;">type</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">99.01</td>
<td style="text-align: left;">category_na</td>
<td style="text-align: left;">na</td>
<td style="text-align: left;">species_match</td>
<td style="text-align: left;">99.99</td>
<td style="text-align: left;">GCA_900458635.1</td>
<td style="text-align: left;">98.99</td>
<td style="text-align: left;">type</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">99.01</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">OK</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">species</td>
<td style="text-align: left;">1290</td>
<td style="text-align: left;">v1.2.2</td>
<td style="text-align: left;">90.97</td>
<td style="text-align: left;">47.945206</td>
<td style="text-align: left;">2.63</td>
<td style="text-align: left;">FDAARGOS_575</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">1290</td>
<td style="text-align: left;">/path/to/genome/GCF_003812505.1/GCF_003812505.1_ASM381250v1_genomic.fna.gz</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Some columns were removed for visual clarity</p>
</div>
<h2 id="taxonomy">Taxonomy</h2>
<p>Table containing the full lineage from kingdom to species of each tax_id</p>
<table>
<thead>
<tr>
<th style="text-align: left;">accession</th>
<th style="text-align: left;">tax_id</th>
<th style="text-align: left;">name</th>
<th style="text-align: left;">rank</th>
<th style="text-align: left;">kingdom</th>
<th style="text-align: left;">phylum</th>
<th style="text-align: left;">class</th>
<th style="text-align: left;">order</th>
<th style="text-align: left;">family</th>
<th style="text-align: left;">genus</th>
<th style="text-align: left;">species</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">GCF_000013425.1</td>
<td style="text-align: left;">93061</td>
<td style="text-align: left;">Staphylococcus aureus subsp. aureus NCTC 8325</td>
<td style="text-align: left;">strain</td>
<td style="text-align: left;">Bacteria</td>
<td style="text-align: left;">Bacillota</td>
<td style="text-align: left;">Bacilli</td>
<td style="text-align: left;">Bacillales</td>
<td style="text-align: left;">Staphylococcaceae</td>
<td style="text-align: left;">Staphylococcus</td>
<td style="text-align: left;">Staphylococcus aureus</td>
</tr>
<tr>
<td style="text-align: left;">GCF_000008865.2</td>
<td style="text-align: left;">386585</td>
<td style="text-align: left;">Escherichia coli O157:H7 str. Sakai</td>
<td style="text-align: left;">strain</td>
<td style="text-align: left;">Bacteria</td>
<td style="text-align: left;">Pseudomonadota</td>
<td style="text-align: left;">Gammaproteobacteria</td>
<td style="text-align: left;">Enterobacterales</td>
<td style="text-align: left;">Enterobacteriaceae</td>
<td style="text-align: left;">Escherichia</td>
<td style="text-align: left;">Escherichia coli</td>
</tr>
<tr>
<td style="text-align: left;">GCF_003812505.1</td>
<td style="text-align: left;">1290</td>
<td style="text-align: left;">Staphylococcus hominis</td>
<td style="text-align: left;">species</td>
<td style="text-align: left;">Bacteria</td>
<td style="text-align: left;">Bacillota</td>
<td style="text-align: left;">Bacilli</td>
<td style="text-align: left;">Bacillales</td>
<td style="text-align: left;">Staphylococcaceae</td>
<td style="text-align: left;">Staphylococcus</td>
<td style="text-align: left;">Staphylococcus hominis</td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.instant", "navigation.instant.prefetch", "content.code.copy"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.50899def.min.js"></script>
      
    
  </body>
</html>